[Android] Parse social pages links in POI editor

strump opened the pull request:
Moved Facebook, Instagram, Twitter, VK cleanup functions from "generator" to "indexer".
Added call to `ValidateAndFormat_*` functions on social links edit.
Fixed `contact:line` validation. Empty string is now correct value.

Closes #1397 
Closes #1546 

## Updated 15.11.2021

Renamed `contacts_processor` to `validate_and_format_contacts`.

Moved functions `ValidateFacebookPage`, `ValidateInstagramPage`, `ValidateTwitterPage`, `ValidateVkPage`, `ValidateLinePage` to `validate_and_format_contacts.hpp/.cpp`

Moved social network tests to separate file `validate_and_format_contacts_test.cpp`. Also added `contact:line` tests there (no more issue #1546 😁 )

Updated XCode project to fix iOS build.

strump commented:
After I moved some functions from "generator" to "indexer" there could be uncertainty in `osm2meta.hpp`:

```cpp
case Metadata::FMD_WEBSITE: valid = ValidateAndFormat_url(v); break;  // <<< here we call method MetadataTagProcessorImpl::ValidateAndFormat_url
case Metadata::FMD_CONTACT_FACEBOOK: valid = ValidateAndFormat_facebook(v); break; // <<< here we call static function
case Metadata::FMD_CONTACT_INSTAGRAM: valid = ValidateAndFormat_instagram(v); break; // <<< here we call static function
case Metadata::FMD_CONTACT_TWITTER: valid = ValidateAndFormat_twitter(v); break; // <<< here we call static function
```

Maybe we should rename `ValidateAndFormat_*` static functions for clarity?

biodranik commented on generator/osm2meta.hpp:
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "indexer/feature_data.hpp"
+#include "indexer/contacts_processor.hpp"
```
We sort includes by alphabet inside one include block.

strump replied:
renamed to `validate_and_format_contacts.hpp` and sorted includes alphabetially.

biodranik commented on indexer/CMakeLists.txt:
```diff
@@ -30,6 +30,8 @@ set(
   complex/serdes_utils.hpp
   complex/tree_node.hpp
+  contacts_processor.cpp
+  contacts_processor.hpp
```
validate_and_format_contacts would be a better name.

strump replied:
renamed to `validate_and_format_contacts`

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+  {
+    if (facebookPage.front() == '@')
+      return facebookPage.substr(1);
+    else
```
This else is redundant.

strump replied:
Fixed

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+      facebookPageUrl = "https://" + facebookPage;
+
+    const url::Url url = url::Url(facebookPageUrl);
+    const string & domain = strings::MakeLowerCase(url.GetWebDomain());
```
1. If a function returns an object, there is no benefit of using const & reference to it (compiler creates and keeps a full string anyway). Just use `string const`.
2. nit: use right-to-left types everywhere, according to our style guide: "a reference to a constant string" is `string const &`

strump replied:
Fixed

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+  {
+    if (instagramPage.front() == '@')
+      return instagramPage.substr(1);
+    else
```
Redundant.

strump replied:
Done!

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+      // Strip last '/' symbol
+      if (webPath.back() == '/')
+        return webPath.substr(0, webPath.length()-1);
+      else
```
Redundant.

strump replied:
Done

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+      // Strip last '/' symbol.
+      if (webPath.back() == '/')
+        return webPath.substr(0, webPath.length()-1);
+      else
```
Redundant.

strump replied:
Fixed

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+  {
+    if (twitterPage.front() == '@')
+      return twitterPage.substr(1);
+    else
```
Redundant.

strump replied:
Fixed

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+    if (vkPageClean.front() == '@')
+      vkPageClean = vkPageClean.substr(1);
+
+    if (vkPageClean.front() == '_' && vkPageClean.back() == '_') return {};
```
nit: easier to debug if the "if's" block (return here) is on the next line. And it is also in our style guide.

strump replied:
Moved return to next line

biodranik commented on indexer/contacts_processor.cpp:
```diff
@@ -0,0 +1,266 @@
+      // Strip last '/' symbol.
+      if (webPath.back() == '/')
+        return webPath.substr(0, webPath.length()-1);
+      else
```
Redundant.

strump replied:
Fixed

biodranik commented on indexer/editable_map_object.cpp:
```diff
@@ -496,27 +497,27 @@ void EditableMapObject::SetWebsite(string website)
 void EditableMapObject::SetFacebookPage(string facebookPage)
 {
-  m_metadata.Set(feature::Metadata::FMD_CONTACT_FACEBOOK, facebookPage);
+  m_metadata.Set(feature::Metadata::FMD_CONTACT_FACEBOOK, ValidateAndFormat_facebook(facebookPage));
```
All these methods take `string const &`. Then you don't need to create a copy of the string in the function argument by using `string facebookPage` (yes, in C++ this declaration creates a copy of the passed parameter). Use `string const & facebookPage` here and in other functions.

strump replied:
Changed argument type

biodranik commented:
1. Curious (without looking at the implementation), why web URL is not validated by the same code used in the generator?
2. There is a build error for iOS, assumingly because of newly added files. Here is the patch to apply (just rename file names if necessary):
```patch
diff --git a/xcode/indexer/indexer.xcodeproj/project.pbxproj b/xcode/indexer/indexer.xcodeproj/project.pbxproj
index 1c6cdfabc1..0399b64158 100644
--- a/xcode/indexer/indexer.xcodeproj/project.pbxproj
+++ b/xcode/indexer/indexer.xcodeproj/project.pbxproj
@@ -184,6 +184,8 @@
                FA67C84626BB356800B33DCA /* categories_cuisines.txt in Resources */ = {isa = PBXBuildFile; fileRef = 4052928E21496D2B00D821F1 /* categories_cuisines.txt */; };
                FA67C84926BB35A400B33DCA /* bounds.hpp in Headers */ = {isa = PBXBuildFile; fileRef = 40C3C090205BF9F400CED188 /* bounds.hpp */; };
                FA67C84F26BB36D700B33DCA /* categories_brands.txt in Resources */ = {isa = PBXBuildFile; fileRef = FA67C84E26BB36D600B33DCA /* categories_brands.txt */; };
+               FA7F9B84273F32680093EA08 /* contacts_processor.hpp in Headers */ = {isa = PBXBuildFile; fileRef = FA7F9B82273F32680093EA08 /* contacts_processor.hpp */; };
+               FA7F9B85273F32680093EA08 /* contacts_processor.cpp in Sources */ = {isa = PBXBuildFile; fileRef = FA7F9B83273F32680093EA08 /* contacts_processor.cpp */; };
                FACB7C0B26B9176500810C9C /* brands_tests.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 406970A121AEF2F10024DDB2 /* brands_tests.cpp */; };
                FACB7C0C26B9176A00810C9C /* centers_table_test.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 3D452AF91EE6D9F5009EAB9B /* centers_table_test.cpp */; };
                FACB7C0D26B9176F00810C9C /* classificator_tests.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 40DF582C2174979200E4E0FC /* classificator_tests.cpp */; };
@@ -416,6 +418,8 @@
                F6DF5F301CD0FD9A00A87154 /* categories_index.hpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.h; path = categories_index.hpp; sourceTree = "<group>"; };
                F6F1DABD1F13D8B4006A69B7 /* ftraits.hpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.h; path = ftraits.hpp; sourceTree = "<group>"; };
                FA67C84E26BB36D600B33DCA /* categories_brands.txt */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text; name = categories_brands.txt; path = ../../data/categories_brands.txt; sourceTree = "<group>"; };
+               FA7F9B82273F32680093EA08 /* contacts_processor.hpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.h; path = contacts_processor.hpp; sourceTree = "<group>"; };
+               FA7F9B83273F32680093EA08 /* contacts_processor.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = contacts_processor.cpp; sourceTree = "<group>"; };
                FACB7C1A26B919AD00810C9C /* libbase.a */ = {isa = PBXFileReference; explicitFileType = archive.ar; path = libbase.a; sourceTree = BUILT_PRODUCTS_DIR; };
                FACB7C1B26B919AD00810C9C /* libcoding.a */ = {isa = PBXFileReference; explicitFileType = archive.ar; path = libcoding.a; sourceTree = BUILT_PRODUCTS_DIR; };
                FACB7C1C26B919AD00810C9C /* libgeometry.a */ = {isa = PBXFileReference; explicitFileType = archive.ar; path = libgeometry.a; sourceTree = BUILT_PRODUCTS_DIR; };
@@ -622,6 +626,8 @@
                                675340AD1A3F540F00A0A8C3 /* classificator_loader.hpp */,
                                675340AE1A3F540F00A0A8C3 /* classificator.cpp */,
                                675340AF1A3F540F00A0A8C3 /* classificator.hpp */,
+                               FA7F9B83273F32680093EA08 /* contacts_processor.cpp */,
+                               FA7F9B82273F32680093EA08 /* contacts_processor.hpp */,
                                34583BC11C88552100F94664 /* cuisines.cpp */,
                                34583BC21C88552100F94664 /* cuisines.hpp */,
                                40D62CEE23F2E8BE009A20F5 /* dat_section_header.hpp */,
@@ -797,6 +803,7 @@
                                6758AED21BB4413000C26E27 /* drules_selector_parser.hpp in Headers */,
                                6753413F1A3F540F00A0A8C3 /* scale_index.hpp in Headers */,
                                6753410E1A3F540F00A0A8C3 /* drawing_rules.hpp in Headers */,
+                               FA7F9B84273F32680093EA08 /* contacts_processor.hpp in Headers */,
                                409EE3E3237E9AA700EA31A4 /* postcodes.hpp in Headers */,
                                670C615C1AB0691900C38A8C /* features_offsets_table.hpp in Headers */,
                                6758AED41BB4413000C26E27 /* drules_selector.hpp in Headers */,
@@ -998,6 +1005,7 @@
                                675341121A3F540F00A0A8C3 /* feature_algo.cpp in Sources */,
                                675341211A3F540F00A0A8C3 /* feature_utils.cpp in Sources */,
                                4099F6491FC7142A002A7B05 /* fake_feature_ids.cpp in Sources */,
+                               FA7F9B85273F32680093EA08 /* contacts_processor.cpp in Sources */,
                                675341231A3F540F00A0A8C3 /* feature_visibility.cpp in Sources */,
                                56C74C221C749E4700B71B9F /* search_delimiters.cpp in Sources */,
                                347F337B1C454242009758CC /* rank_table.cpp in Sources */,
```

strump commented:
Moved regex's to static variables to reuse same regular expressions in validation and formatting functions. Now all regexes could be found at first lines of `validate_and_format_contacts.cpp`.
Looks like I'm done with this PR 🕺🏻

biodranik commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+  if (regex_match(page, s_instaRegex))
+    return true;
+
+  if (EditableMapObject::ValidateWebsite(page))
```
nit: Negate the if, return false, and make the code shorter without bracers.

strump replied:
Done

vng commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+    // Check Twitter domain name.
+    if (domain == "twitter.com" || strings::EndsWith(domain, ".twitter.com"))
+    {
+      auto webPath = url.GetWebPath();
```
Э, как-то сложно. Судя по логике, pop_back и pop_front будет просто и достаточно, ну и конечно надо проверить на empty предварительно.

biodranik replied:
Тогда на empty придётся проверять дважды. Это самый простой способ, он ещё кстати обработает @@@@

strump replied:
Get rid of `.back()` method using your advice

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      return false;
+
+    if (regex_match(vkLogin, s_badVkRegex))
+      return false;
```
Simplified if() statement. But also added parentheses `vkLogin.front() == '_' && vkLogin.back() == '_'`

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+    {
+      auto webPath = url.GetWebPath();
+      // Strip last '/' symbol
+      if (webPath.back() == '/')
```
Get rid of `.back()` method using your advice

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+    {
+      auto webPath = url.GetWebPath();
+      // Strip last '/' symbol.
+      if (webPath.back() == '/')
```
Get rid of `.back()` method using your advice

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      return twitterPage.substr(1);
+    return twitterPage;
+  }
+  if (EditableMapObject::ValidateWebsite(twitterPage))
```
Negated if() statement

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+    if (vkPageClean.front() == '_' && vkPageClean.back() == '_')
+      return {};
+    if (regex_match(vkPageClean, s_badVkRegex))
+      return {};
```
Simplified if() statement. But also added parentheses `(vkLogin.front() == '_' && vkLogin.back() == '_')`

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+            domain.find(".facebook.") != string::npos || domain.find(".fb.") != string::npos);
+  }
+
+  return false;
```
Simplified this block using negated `if (!EditableMapObject::ValidateWebsite(page))`

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      linePageUrl = "https://" + page;
+
+    const url::Url url = url::Url(linePageUrl);
+    const string &domain = strings::MakeLowerCase(url.GetWebDomain());
```
Done

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+    if (!strings::StartsWith(page, "http://") && !strings::StartsWith(page, "https://"))
+      linePageUrl = "https://" + page;
+
+    const url::Url url = url::Url(linePageUrl);
```
Done

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+  if (EditableMapObject::ValidateWebsite(page))
+  {
+    string linePageUrl = page;
+    // Check if HTTP protocol is present
```
Done

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+  return false;
+}
+
+}
```
Done

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      linePageUrl = "https://" + linePage;
+
+    url::Url const url = url::Url(linePageUrl);
+    string const & domain = strings::MakeLowerCase(url.GetWebDomain());
```
Done

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      vkPageUrl = "https://" + vkPage;
+
+    const url::Url url = url::Url(vkPageUrl);
+    const string & domain = strings::MakeLowerCase(url.GetWebDomain());
```
Updated type declaration

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      twitterPageUrl = "https://" + twitterPage;
+
+    const url::Url url = url::Url(twitterPageUrl);
+    const string & domain = strings::MakeLowerCase(url.GetWebDomain());
```
Done (+ const)

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      instagramPageUrl = "https://" + instagramPage;
+
+    const url::Url url = url::Url(instagramPageUrl);
+    const string & domain = strings::MakeLowerCase(url.GetWebDomain());
```
Done (+ const)

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      return instagramPage.substr(1);
+    return instagramPage;
+  }
+  if (EditableMapObject::ValidateWebsite(instagramPage))
```
Nedated if() statement

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      return facebookPage.substr(1);
+    return facebookPage;
+  }
+  if (EditableMapObject::ValidateWebsite(facebookPage))
```
Done

strump commented on indexer/validate_and_format_contacts.cpp:
```diff
@@ -0,0 +1,400 @@
+      // Strip last '/' symbol
+      if (webPath.back() == '/')
+        return webPath.substr(0, webPath.length()-1);
+      else
```
Fixed

strump commented:
It appeared that `Url::FromString` method handles HTTP(S) protocol correctly. So I removed all checks for `"http://"` and `"https://"` in `validate_and_format_contacts.cpp` code and use only `Url::FromString(...)`. Also fixed comments and const variable definitions.

biodranik commented:
Current implementation of the Url class has some bugs now. For example, this works:
```c++
Url u = Url::FromString("https://");
ASSERT(u.IsValid(), ());
//
u = Url::FromString("@&€:1;asf");
ASSERT(u.IsValid(€, ());
```

strump commented:
@biodranik in `validate_and_format_contact.cpp` code I use `Url::FromString(page)` only after calling `EditableMapObject::ValidateWebsite(page)`. We can add invalid urls to unit-test

vng commented:
I don't mind to merge. Url::FromString is not changed in this PR and it's called after url validation. We don't break any constrains comparing to the master branch. If we want some refactoring, let's make separate PR, otherwise we will stuck here for a long time.

biodranik commented:
Thank you for your huge work and time! I hope you learned something new on the way :)
